### PR TITLE
Nitrokey v0.8.0

### PR DESCRIPTION
**Pull request by robinkrahl**
This is a preparation for the upcoming release of nitrokey-rs v0.8.0.

Open tasks:
- The `Display` implementation for `nitrokey::Model` changed.  We should check where we use it and if we have to change anything.
- We should change `DeviceModel::as_user_facing_str` to print the full model name.  And we should make sure that it is actually used -- a quick grep didn’t show any matches.
- We have to update the man page for the Librem model.

Questions:
- We should revisit the option names for `config set` (`--{no-,}numlock`, `--{no-,}capslock`, `--{no,}scrollock`).  The names `numlock`, `capslock`, `scrolllock` (three L) originate in libnitrokey.  Due to a typo, `scrolllock` became `scrollock` in nitrokey-rs, and I didn’t notice the mistake when implementing the command for nitrocli.  The next release will fix this typo and rename the fields to `num_lock`, `caps_lock` and `scroll_lock` to make them easier to read and to make them match the terms Num Lock, Caps Lock, Scroll Lock.  We should either rename the nitrocli options to `--{no-,}num-lock` etc. (preferred) or at least fix the `scrolllock` typo.  What do you think?

### CODE REVIEW

**Comment by d-e-s-o** [comment]
> Questions:
> 
> * We should revisit the option names for `config set` (`--{no-,}numlock`, `--{no-,}capslock`, `--{no,}scrollock`).  The names `numlock`, `capslock`, `scrolllock` (three L) originate in libnitrokey.  Due to a typo, `scrolllock` became `scrollock` in nitrokey-rs, and I didn’t notice the mistake when implementing the command for nitrocli.  The next release will fix this typo and rename the fields to `num_lock`, `caps_lock` and `scroll_lock` to make them easier to read and to make them match the terms Num Lock, Caps Lock, Scroll Lock.  We should either rename the nitrocli options to `--{no-,}num-lock` etc. (preferred) or at least fix the `scrolllock` typo.  What do you think?

Isn't the three L thing more of a (new) German peculiarity? Don't get me wrong, it makes sense, I just don't recall ever having seen that it English. In any case, I think inserting a hyphen as you suggest makes sense, so we would circumvent the issue altogether.

I didn't have the cycles to take a look at this pull request today. I hope to get to it tomorrow or over the weekend the latest.

**Comment by robinkrahl** [comment]
> Isn't the three L thing more of a (new) German peculiarity?

I’m not aware of any rules for omitting repeated letters in English, but I think that’s mainly because you typically don’t have this problem if you separate compound words.  I guess the English solution is to add a hyphen or space, as in [ill-looking](https://www.merriam-webster.com/dictionary/ill-looking), just as we do here.

**Comment by d-e-s-o** [comment]
Looks good to me so far. Thanks for working on this bit! Are you still ironing out details for nitrokey-rs 0.8.0 itself?

**Comment by robinkrahl** [comment]
Thanks for the review!  I was waiting for an update of the Arch
libnitrokey package so that I can also run the unit tests against a
precompiled libnitrokey version.  The package has been updated a few
days ago, so I should be able to test and publish nitrokey-rs 0.8.0
today or tomorrow.

**Comment by robinkrahl** [comment]
I’ve released nitrokey-rs v0.8.0 and updated the PR.  It should be ready now.  Could you please have another look?

**Comment by d-e-s-o** [comment]
Awesome! Thank you! I hope to get to that in the next few days.

File: src/commands.rs [hunk 1]
```diff
@@ -667,13 +657,13 @@ pub fn config_get(ctx: &mut Context<'_>) -> anyhow::Result<()> {
     println!(
       ctx,
       r#"Config:
-  numlock binding:          {nl}
-  capslock binding:         {cl}
-  scrollock binding:        {sl}
+  num lock binding:         {nl}
+  caps lock binding:        {cl}
+  scroll lock binding:      {sl}
```
Comment (d-e-s-o) [inline src/commands.rs:662]: Would it make sense to move these changes into the "Rename numlock, capslock, scrollock options" commit?

**Review by d-e-s-o — COMMENTED** [review]
Sorry for the delay. Thanks for the adjustments. Looks good! I guess before we merge this we'll need to get `nitrokey-test` up to speed. I'll see if I can get this done over the weekend.

**Comment by d-e-s-o** [comment]
@insanerwayner you can use Robin's branch if you want to have Librem Key support and don't mind cloning directly instead of pulling from `crates.io`. We welcome beta testers!

**Comment by robinkrahl** [comment]
Thanks for the review!

> I guess before we merge this we'll need to get nitrokey-test up to speed.

In theory yes, but does it really matter if we don’t have a test device?

File: doc/nitrocli.1 [hunk 2]
```diff
@@ -245,8 +245,8 @@ If \fB\-\-no\-otp\-pin\fR is set, OTP generation can be performed without PIN.
 These two options are mutually exclusive.
 
 .SS Password safe
-The Nitrokey Pro and the Nitrokey Storage provide a password safe (PWS) with 20
-slots.
+The Nitrokey Pro, the Nitrokey Storage, and the Librem Key provide a password
+safe (PWS) with 20 slots.
```
[the remark this answers is not in the document]
Comment (robinkrahl) [reply doc/nitrocli.1:249]: I think it should be 16 for all devices, see also the [`nitrokey::SLOT_COUNT`](https://docs.rs/nitrokey/0.8.0/nitrokey/constant.SLOT_COUNT.html) constant.  I don’t know where the 20 came from.

File: src/tests/run.rs [hunk 3]
```diff
@@ -171,34 +172,37 @@ fn connect_wrong_usb_path(_model: nitrokey::Model) {
 
 #[test_device]
 fn connect_model(_model: nitrokey::Model) -> anyhow::Result<()> {
+  use std::convert::TryInto;
```
[the remark this answers is not in the document]
Comment (robinkrahl) [reply src/tests/run.rs:175]: I generally try to import into the smallest required scope.  In my opinion, this makes it easier to read and understand the imports, especially for modules and traits that are not very common.  That’s not that big of an issue for the standard library, so I don’t have any objections against moving this import to the beginning of the file.

**Comment by robinkrahl** [comment]
Updated to address the comments.

**Comment by d-e-s-o** [comment]
> Thanks for the review!
> 
> > I guess before we merge this we'll need to get nitrokey-test up to speed.
> 
> In theory yes, but does it really matter if we don’t have a test device?

Perhaps @jans23 is willing to provide us with some?

Jan, is there any chance that the Nitrokey GmbH can provide Robin (assuming he wants/needs) and me with a Librem Key? We would like to support the device on the [`nitrokey-rs`](https://crates.io/crates/nitrokey) and `nitrocli` side of things, but our inability to test specifically with it makes that more random guessing than engineering. Given that it is reported differently from a Nitrokey Pro just testing with the latter is not sufficient.

**Comment by szszszsz** [comment]
We would be happy to, but I believe we do not have the actual firmware image for Librem Key. From what I know the only difference is the VID:PID pair, but this might be outdated.

**Comment by d-e-s-o** [comment]
> We would be happy to, but I believe we do not have the actual firmware image for Librem Key. From what I know the only difference is the VID:PID pair, but this might be outdated.
> Edit: Librem Key source code links to Nitrokey Pro firmware

Thanks for your comment. So in that case, could you provide a regular Pro that reports the different VID:PID pair? How is `libnitrokey` tested for these devices?

**Comment by szszszsz** [comment]
1. I think better solution would be to ask @purism directly to cover their hidden firmware changes and future updates as well.
2. We do not test Librem Keys, but asking volunteers to do so (simple connection will do). As long as the VID:PID is the only change, there is no need to cover it more I believe.

File: src/tests/run.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (robinkrahl) [reply src/tests/run.rs:175]: I just found [this blog post](https://tarquin-the-brave.github.io/blog/posts/rust_use_statements/) via This Week in Rust that also discusses this topic.

**Comment by d-e-s-o** [comment]
I've merged the pull request to not have it linger here for too long. I am not sure when I will be able to update `nitrokey-test` to work with the Librem Key. Robin, if you are interested in this work, feel free to take a look.

**Pull request closed without merging**